sapicore/engine/component/__init__.py

```diff
@@ -0,0 +1,183 @@
+"""Components handle universal parameters and logic common to neurons, synapses, and their derivatives."""
+from typing import Any
+import os
+
+import torch
+from torch import Tensor
+from torch.nn import Module
+
+from sapicore.utils.constants import DT
+from sapicore.utils.loggable import Loggable
+from sapicore.utils.sweep import Sweep
+from sapicore.utils.io import save_yaml
+
+from tree_config import Configurable, apply_config
+
+__all__ = ("Component",)
+
+
+class Component(Module, Configurable, Loggable):
+    """Model component base class.
+
+    Defines instance attributes and methods shared by all model components.
+    Its primary purpose is to implement generic operations on loggable and configurable attributes.
+
+    Parameters
+    ----------
+    identifier: str, optional
+        Human-readable identifier for the component instance.
+
+    configuration: dict, optional
+        Configuration dictionary used during parameter initialization.
+
+    device: str, optional
+        Specifies a hardware device on which to process this object's tensors, e.g. "cuda:0".
+        Must be a valid torch.device() string value. Defaults to "cpu".
+
+    kwargs:
+        Additional instance attributes that the user may set.
+
+    See Also
+    --------
+    :class:`~tree_config.Configurable`
+    :class:`~utils.loggable.Loggable`
+    :class:`~utils.sweep.Sweep`
+
+    """
+
+    def __init__(self, identifier: str = None, configuration: dict = None, device: str = "cpu", **kwargs):
+        super().__init__()
+
+        # model-related common instance attributes.
+        self.identifier = identifier
+        self.configuration = configuration
+
+        # hardware device on which to process this component.
+        self.device = device
+
+        # simulation-related tracking variables.
+        self.simulation_step = 0
+        self.dt = DT
+
+        # initialize configurable and loggable attributes with dummy tensors.
+        for prop in self._config_props_:
```

Comment:
`Loggable` and `Configurable` are used here in a way that doesn't match the scenario you had in mind when you designed it (that's why I was reluctant to use it initially--I needed something **like** it, but not exactly it).

The snippet here would raise an exception if you were to iterate over `_config_props` (missing attribute). Try running simulation.py in debug mode, and you'll see that network components (ensembles, etc.) are built incrementally through a chain of super() calls. `Component` is the end of the line, and what it does here is create these attributes for the first time (since there's no convenient/intuitive way to do it with the loggable API--recall our back and forth on slack regarding the behavior of `apply_config`).

Reply:
Hmm I think from my conception, whatever class variables exist or used, they should be created as such (i.e. class variable ideally IMO - although I know you don't like that or instance variables in init before super is called). Because ideally these variables should exist and not created from the config files. Config files should only set existing variables and not create them.

If you follow this approach would there still be an issue and exceptions? Or is there a reason to not make sure variables exist first?

Reply:
> These variables should exist and not created from the config files.

The attributes aren't created from config files, but from the `_config_props_` variable defined before `__init__`. `Component.__init__` iterates over `_config_props_` because, by design, it doesn't know those variables for each child object that might use it in a super call. Since it is agreed that every child class of `Configurable` has `_config_props_`, attribute initialization can be deferred to the parent `Component` constructor (iterating over that tuple), preventing code duplication. That's why it's there.

To clarify: every time you initialize a complex object like an Ensemble, you start a cascade of super calls where `Component` is the top ancestor. Each super call adds attributes shared by all of its children (Neuron's constructor would add `equation`, SpikingNeuron's constructor would add the `spiked` buffer, etc.). Importantly, `_configurable_props_` are **not** used as a catch-all capturing every possible descriptive attribute.

P.S. I moved the `input` and `voltage` declarations in `Neuron` into the constructor for consistency with the above.